### DISCUSSION TITLE
Best way to align celestial objects with a landscape for photography?

### DISCUSSION THREAD

**Discussion by alphanimal** [Stellarium/stellarium#2003] [Ideas]
Hi!

As a photographer, I have found many locations with a nice landscape or a subject in the distance, with the sky in the background. It would be very useful to see how celestial objects will line up in that photo at different points in time, to plan for photos with the sun, moon, a constellation or a nebula etc. in the background.

To plan for such photos, you need the horizontal (alt/az) coordinates of your subject. You can get that from apps like Sun Surveyor, by placing markers representing your camera and subject on a map and the app calculates the coordinates. Most of the times the altitude is not very accurate, because the map's elevation data isn't precise and it's hard to get the exact elevation of vertical features of mountains, buildings etc.

To get more precise coordinates, it's best to look at a photo from that location at a known time, with celestial objects visible. You can then set Stellarium to the same location and time as the photo, and place a marker (Shift-click) on the part in the sky where your subject is on the photo. You can then read the alt/az coordinates from that marker.

Once you have the coordinates of your subject, you can adjust your date and time in Stellarium to see when you want to go out and take a photo with your favorite object in the background. You can even use the Oculars feature in Stellarium to visualize the framing of the photo (add a sensor and telescope matching the camera's sensor size and the lenses focal length)

The problem is, the markers you place with Shift-click are fixed to equatorial coordinates, so they move to different locations in the sky when you change time. One way I found to get around this is to open the scripting console (F12) and use a command to place a marker on the horizontal coordinate system, which doesn't move relative to the landscape.

This will place a marker at azimuth 180°, altitude 10° with a specific shape, color and size:

```
MarkerMgr.deleteAllMarkers();
MarkerMgr.markerHorizon(180, 10, true, "dashed-square", "#ffff80", 20);
```

I found this very useful so far. You can just save a script (.scc file) with markers for all your locations and open that to start Stellarium with everything marked. One problem here is, I couldn't find a way to add labels to those markers, to know which location is which.

I'm thinking it would be awesome to have a feature like this built into Stellarium with a GUI. There's already the "Observing lists" feature, where you can save lists of objects, including custom markers set with Shift-click.
The markers could be adapted to allow saving horizontal coordinates instead of only equatorial coordinates.

The ultimate feature would be loading and aligning photos directly in Stellarium. I imagine you could just open a photo in Stellarium and it will set the time and location from the image metadata (or you enter it manually). Then you can adjust the image's position, scale, rotation to line up the photos' sky with Stellarium's sky. Maybe even allow to correct for perspective and lens distortions, which would be required for wide angle photos, but photos with a small FOV should be accurate enough without corrections. Then you edit that photo externally to remove the sky, and load it back into Stellarium, to make a pretty good visualization of how that photo would look at any time of the night or day.

I've seen there's a way to load images from scripts using `core.loadSkyImage()`, but I haven't tried that yet, and I think those images would have to be aligned by entering numbers in the script, without a GUI.

I'd appreciate any thoughts or tips on that subject! Thanks!

**Comment by gzotti** [comment]
You can create an artificial pano following the User Guide section 7.3.4 (or some other methods described elsewhere), then try to align your daylight photos with that and the Hugin stitcher. Else please consider https://github.com/Stellarium/stellarium/wiki/FAQ#why-dont-you-implement as answer.

**Comment by alphanimal** [comment]
Thanks for the reply!

I've now read the "Why don't you implement..." section in the FAQ. Sorry if it came across like that, but I'm not trying to demand any features. If you know of any other software that has this feature I'd be interested too :) I'm just throwing out ideas and asking if there might be a better way to do things, that I'm unaware of. After I figured out that you can create markers in the horizontal coordinate system using scripting, that already made this task a lot easier for me. Also I want to share my finding so others might benefit from it.

I have looked into creating custom landscapes in the past too! I might read the user guide on that topic to get some more insight, but from what I remember it involves lots of manual work to precisely align image in a landscape.

Disregarding actually loading/aligning photos inside Stellarium, which I understand would be a huge amount of work, I think expanding the abilities of markers would benefit a lot of users. Just being able to set markers to be in horizontal coords instead of equatorial would be so useful! And to be able to change the label and save them in a list as well.

Thanks for reading my suggestions!

**Reply by gzotti** [reply]
Alright, maybe the markers in horizontal coordinates. But the image alignment really is nothing for Stellarium.

**Reply by alphanimal** [reply]
I understand that - it would be quite a big feature. I think permanent AzAlt markers with custom labels would be very useful though. I'm trying to get the `core.loadSkyImage` scripting function to work right now but I'm stuck. Where would be the best place to ask scripting questions? Thanks!

edit: I found your test scripts for loadSkyImage, and those work! I'll figure it out :)
edit 2: It was 3 things:
- I had magnitude set to the default 14 which is almost invisible. -14 works better
- the azimuth coordinates seem to be backwards. you have to calculate 180-x to get from the value visible in GUI to the value you need for scripting.
- The path for the image file didn't resolve correctly (I still haven't figured out how to specify absolute paths on Windows, I have only managed to get it to with with relative paths)

**Comment by alphanimal** [comment]
I have managed to get `core.loadSkyImage()` working now! This means I can actually load photos right into Stellarium and they stay fixed to the AzAlt frame! Amazing!

I still have to do some calculations and alignment without a GUI, by entering numbers into the script, but I can actually use this! I'm sure I can make a spreadsheet or something to help with the alignment.

**Reply by HenryWede** [reply]
Hello alphanimal - 
I know this post is a year old, but would you be willing to post an image or two of how you did this?  It would be very helpful to have a starting point as I wander through the documentation.
Thank you in advance!

**Reply by alphanimal** [reply]
Hi! I have actually used that script for a few shots already, but before I share it I want to clean it up a little

**Reply by HenryWede** [reply]
That sounds great - thank you for your time.  I think this will lead me completely out of my comfort zone, but at least I'll learn something new.
